tests/43search.pl

```diff
@@ -53,3 +54,109 @@
          Future->done(1);
       });
    };
+
+test "Can back-paginate search results",
+    requires => [ local_user_and_room_fixtures() ],
+
+    check => sub {
+        my ( $user, $room_id ) = @_;
+
+        my ( @event_ids );
+
+        my $search_query = {
+            search_categories => {
+                room_events => {
+                    keys => [ "content.body" ],
+                    search_term => "Message",
+                    order_by => "recent",
+                }
+            }
+        };
+
+        repeat( sub {
+            my $msgnum = $_[0];
+            matrix_send_room_text_message( $user, $room_id,
+                                           body => "Message number $msgnum" )
+                ->on_done( sub { ($event_ids[$msgnum]) = @_ } )
+        }, foreach => [ 0 .. 19 ] ) -> then ( sub {
+            do_request_json_for( $user,
+                                 method  => "POST",
+                                 uri     => "/api/v1/search",
+                                 content => $search_query,
+            );
+        })->then( sub {
+            my ( $body ) = @_;
+
+            log_if_fail "First search result:", $body;
+
+            assert_json_keys( $body, qw( search_categories ) );
+            assert_json_keys( $body->{search_categories}, qw ( room_events ) );
+            my $room_events = $body->{search_categories}{room_events};
+
+            assert_json_keys( $room_events, qw( count results next_batch ) );
+
+            $room_events->{count} == 20 or die "Expected count == 20";
+
+            my $results = $room_events->{results};
+            scalar @$results == 10 or die "Expected 10 search results";
+            $results->[0]{result}{event_id} eq $event_ids[19]
```

Comment:
These can be done with

    assert_eq( $results->[0]{result}{event_id}, $event_ids[19],
       'first result event_id' );

which prints the expected/actual values on failure